File: Sharebook/Client/src/app/core/Core/Extensions/StringFormater.ts
import StringBuilder from "../../../c#-objects/TypeScript.NET-Core/packages/Core/source/Text/StringBuilder";

export class StringFormatter {
  private static readonly UNRESERVED_CHARS: string = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_.~";

  // Clean a string so that it can be used in an URL
  // <param name="str">string to clean</param>
  // <returns>Cleaned string that can be added into an URL</returns>
  public static urlEncode(str: string): string {
    let result: StringBuilder = new StringBuilder();

    for (let char of str) {
      if (StringFormatter.UNRESERVED_CHARS.includes(char.toString())) {
        result.append(char);
      } else {
        result.append('%' + `${StringFormatter.ascii_to_hexadecimal(char)}`);
      }
    }

    return result.toString();
  }

  private static ascii_to_hexadecimal(str: string) {
    let arr1 = [];
    for (let n = 0, l = str.length; n < l; n++) {
      let hex = Number(str.charCodeAt(n)).toString(16);
      arr1.push(hex);
    }

    return arr1.join('');
  }
}

File: Server/src/Chessbook.Common/Extensions/StringFormater.cs
﻿namespace Chessbook.Common.Extensions
{
    using System.Text;

    /// <summary>
    /// Class providing methods to format a string
    /// </summary>
    public class StringFormater
    {
        private const string UNRESERVED_CHARS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_.~";

        /// <summary>
        /// Clean a string so that it can be used in an URL
        /// </summary>
        /// <param name="str">string to clean</param>
        /// <returns>Cleaned string that can be added into an URL</returns>
        public static string UrlEncode(string str)
        {
            StringBuilder result = new StringBuilder();

            foreach (char c in str)
            {
                if (UNRESERVED_CHARS.Contains(c.ToString()))
                {
                    result.Append(c);
                }
                else
                {
                    result.Append('%' + $"{(int) c:X2}");
                }
            }

            return result.ToString();
        }
    }
}
